objects/adffc0/contained/706d9f/contained/4363a8/script.lua

```diff
@@ -0,0 +1,299 @@
+-- Gives the object a tag to indicate it's submerged in a board.
+function populateTile(bag, guid, position, rotation)
+    -- TODO: Scale this properly according to scaleFactors[SetupChecker.getVar("optionalScaleBoard")].size
+    -- Get the object out of the bag to clone it, but use a callback to put it back as soon as it spawns.
+    local original = bag.takeObject({
+        guid = guid,
+        position = bag.getPosition() + Vector(0, 10, 0),
+        callback_function = function(obj) bag.putObject(obj) end,
+    })
+    -- Spawn the clones somewhere out of the way, so we can lock them before moving them into position.
+    local clone = original.clone({position = original.getPosition() + Vector(0, 10, 0)})
+    clone.setLock(true)
+    clone.setPosition(position)
+    clone.setRotation(rotation)
+    clone.addTag(preparedTag)
+    clone.interactable = false
+    clone.addTag("Uninteractable")
+    Wait.condition(
+        function() toPopulate = toPopulate - 1 end,
+        function() return not clone.spawning and not clone.loading_custom end
+    )
+    return clone.getGUID()
+end
+
+-- Calculates the correct position for a tile, given a board.
+function calculatePosition(board, tilePosition)
+    local boardScale = board.getScale()
+    local localPosition = Vector(-tilePosition.x / boardScale.x, preparedTileHeight / boardScale.y, -tilePosition.z / boardScale.z)
+    return board.positionToWorld(localPosition)
+end
+
+-- Takes a board and fills it with drowning tiles, locked and hidden inside it.
+function populateBoard(board)
+    if checkBoard(board) then
+        return
+    end
+    
+    -- Determine the appropriate bag.
+    local bag
+    local positions
+    if board.hasTag("Balanced") then
+        bag = getObjectFromGUID(balancedBag)
+        positions = balancedPositions
+    else
+        bag = getObjectFromGUID(thematicBag)
+        positions = thematicPositions
+    end
+    
+    -- Go through the bag, storing the GUIDs of the objects we want.
+    -- We store GUIDs as populateTile modifies the bag contents, which may interfere with bag operations.
+    local guids = {}
+    for _,obj in pairs(bag.getObjects()) do
+        if string.find(obj.name, "^"..board.getName().."[0-9]*$") then
+            guids[obj.name] = obj.guid
+            toPopulate = toPopulate + 1
+        end
+    end
+    
+    local drowningTiles = {}
+    -- Copy all the appropriate objects into place.
+    for name,guid in pairs(guids) do
+        local spawnedGUID = populateTile(bag, guid, calculatePosition(board, positions[name]), board.getRotation())
+        drowningTiles[name] = spawnedGUID
+    end
+    
+    markBoard(board, drowningTiles)
+    startCastDownTimer()
+end
+
+-- Adds right-click functionality to a tile to undrown itself.
+function addUndrown(tile)
+    -- We also need to ensure it will restore this functionality to itself if reloaded.
+    tile.setLuaScript(string.format([[
```

Comment:
does `setLuaScript` automatically reload the object or should we also call `tile.reload()`?